[Follow-Up] Updated TC_ACL_2_7 and TC_ACL_2_8 python3 test module to include fabric sensitive checks

j-ororke opened the pull request:
#### Testing
Updated TC_ACL_2_7 and TC_ACL_2_8 python3 test modules:
- Updated to conform these 2 test modules to the follow-up task here: [549](https://github.com/project-chip/matter-test-scripts/issues/549)
- Added fabric unfiltered checks to test steps 7-10 for TC_ACL_2_7 and TC_ACL_2_8 to make sure that with fabric sensitive reads that the writes don't pass over to other fabrics data frames
- Test Plan PR Link: https://github.com/CHIP-Specifications/chip-test-plans/pull/5139

j-ororke committed:
Updated TC_ACL_2_7 and TC_ACL_2_8 python3 test module:
- Updated to conform these 2 test modules to the follow-up task here: https://github.com/project-chip/matter-test-scripts/issues/549
- Added fabric_filtered/fabricFiltered args to test steps 7-10 for TC_ACL_2_7 and TC_ACL_2_8 to include these checks to make sure that with fabric sensitive reads that the writes dont pass over to other fabrics data

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_7-8_fabrics-sensitive

cecille commented on src/python_testing/TC_ACL_2_8.py:
```diff
@@ -73,6 +73,74 @@ def _verify_acl_event(
+        asserts.assert_equal(found_valid_events, expected_events_count,
+                             f"Expected {expected_events_count} valid events for TH1, found {found_valid_events}")
+
+        if is_filtered:
```
Are you getting back events for TH2 when you don't use fabric filtering? These are fabric sensitive, so they shouldn't be emitted either way....

j-ororke replied:
Hi Cecille,
So yes when not using fabric filtered we get back results for TH1 when filtering with TH2, it appears that the data section is just obscured or filtered out for the extensions attribute, below is an example from test step 8 output for TH2 with fabric filtered = false:
`[AccessControl.Structs.AccessControlExtensionStruct(data=b'', fabricIndex=1), AccessControl.Structs.AccessControlExtensionStruct(data=b'\x17\xd0\x00\x00\xf1\xff\x01\x00=Hello World. This is a single element living as a charstring\x00\x18', fabricIndex=2)]`

j-ororke replied:
Sorry, I see the issue here now, the above was from ACL_2_7 test step 8:
[AccessControl.Structs.AccessControlExtensionStruct(data=b'', fabricIndex=1), AccessControl.Structs.AccessControlExtensionStruct(data=b'\x17\xd0\x00\x00\xf1\xff\x01\x00=Hello World. This is a single element living as a charstring\x00\x18', fabricIndex=2)]


The result we get from ACL_2_8 test step 8 when TH2 reads ACL result (fabric_filtered=True): [AccessControl.Structs.AccessControlEntryStruct(privilege=<AccessControlEntryPrivilegeEnum.kAdminister: 5>, authMode=<AccessControlEntryAuthModeEnum.kCase: 2>, subjects=[2, 2222], targets=Null, fabricIndex=2)]

However, we get the following during ACL_2_8 test step 8 when TH2 reads ACL result (fabric_filtered=False): [AccessControl.Structs.AccessControlEntryStruct(privilege=0, authMode=0, subjects=Null, targets=Null, fabricIndex=1), AccessControl.Structs.AccessControlEntryStruct(privilege=<AccessControlEntryPrivilegeEnum.kAdminister: 5>, authMode=<AccessControlEntryAuthModeEnum.kCase: 2>, subjects=[2, 2222], targets=Null, fabricIndex=2)]

j-ororke replied:
Additionally, regarding the data output b'\x17\xd0\x00\x00\xf1\xff\x01\x00=Hello World. This is a single element living as a charstring\x00\x18' this came from ACL_2_7 test step 8, I investigated this further and discussed it with Amine to clarify what's happening here just to make certain that this behavior is correct and expected. The data is coming over the wire exactly as it should be; it's not being substituted or modified by the client. Here's a breakdown of what is occurring:

*The D_OK_SINGLE constant represents valid TLV (Tag-Length-Value) encoded data that containing:*
- TLV structure bytes: \x17\xd0\x00\x00\xf1\xff\x01\x00=
- An embedded character string element: Hello World. This is a single element living as a charstring
- TLV termination bytes: \x00\x18

When this TLV-encoded payload is transmitted and received, we see both the binary TLV structure and the embedded string content in its human-readable form. This is the standard behavior for TLV data that contains string elements - the string content remains visible within the binary representation.
The ACL_2_7 test step 8 is functioning correctly, and this output confirms that the TLV encoding/decoding is working as designed. I apologize for having provided the incorrect data to this comment previously.

j-ororke replied:
As discussed in Slack after you have reviewed the trace log output from ACL_2_7 python3 test module test run for test steps 7 and 8, added the additional checks to make sure that data was empty for other fabrics in unfiltered response:
 - Added check to make sure that TH1 is empty binary data in response if TH2 is reading extensions unfiltered 
 - Added check to make sure that TH2 has empty binary data in response if TH1 is reading extension data unfiltered

cecille replied:
OK - just to double check - 'cause I want to be sure before i start a fire...

You're getting fabric sensitive events back for the non-accessing fabric, but there's no data inside. But you're still getting the events.

This still feels like an implementation bug that I'd like to chase down. Because it's still leaking information about the other fabric's ACL management.

cecille commented on src/python_testing/TC_ACL_2_7.py:
```diff
@@ -148,128 +202,98 @@ async def test_TC_ACL_2_7(self):
+            fabric_filtered=False
+        )
+        logging.info("TH2 read result (fabricFiltered=False): %s", str(result2_unfiltered))
+        asserts.assert_equal(len(result2_unfiltered), 2, "Should have two extensions when not fabric filtered")
```
two things
1) you don't know for sure that there's no other fabric on the DUT, so this might not actually be two. There should be at least two
2) The important check here isn't so much about non-fabric filtered reads returning data. The data should be returned, yes, but the fabric sensitive fields in the struct aren't supposed to be readable from other fabrics. So if you read here with fabric_filtered = False, you should get back an entry for TH2, but it shouldn't tell you anything interesting. I don't love the spec wording here - it's not nearly specific enough. The subject and target are nullable, but I'd actually be quite interested to know what happens with the privilege and auth mode fields. Actually, what ARE you getting back here for the

j-ororke replied:
Hi Cecille,
Understood, I have updated this part for the 1st point to:
- asserts.assert_true(len(result2_unfiltered) >= 2, "Should have two extensions when not fabric filtered")
Additionally, I have provided the output from this test step here:
[AccessControl.Structs.AccessControlExtensionStruct(data=b'', fabricIndex=1), AccessControl.Structs.AccessControlExtensionStruct(data=b'\x17\xd0\x00\x00\xf1\xff\x01\x00=Hello World. This is a single element living as a charstring\x00\x18', fabricIndex=2)]

However, I think this might be a bit more interesting regarding what you had mentioned for point 2 and auth fabric filtering, when we read the AccessControlList attribute with TH1 and fabric filtered = False: [AccessControl.Structs.AccessControlEntryStruct(privilege=<AccessControlEntryPrivilegeEnum.kAdminister: 5>, authMode=<AccessControlEntryAuthModeEnum.kCase: 2>, subjects=[112233, 1111], targets=Null, fabricIndex=1), AccessControl.Structs.AccessControlEntryStruct(privilege=0, authMode=0, subjects=Null, targets=Null, fabricIndex=2)]

Then If we look and compare to the result of the AccessControlList attribute for TH1 with fabric filtered = True:
[AccessControl.Structs.AccessControlEntryStruct(privilege=<AccessControlEntryPrivilegeEnum.kAdminister: 5>, authMode=<AccessControlEntryAuthModeEnum.kCase: 2>, subjects=[112233, 1111], targets=Null, fabricIndex=1)]

j-ororke replied:
As discussed in Slack after you have reviewed the trace log output from ACL_2_7 python3 test module test run for test steps 7 and 8, added the additional checks to make sure that data was empty for other fabrics in unfiltered response:

- Added check to make sure that TH1 is empty binary data in response if TH2 is reading extensions unfiltered
- Added check to make sure that TH2 has empty binary data in response if TH1 is reading extension data unfiltered

j-ororke committed:
Updating TC_ACL_2_7 python3 test module:
- changed from equal to a greater than value to verify that we got more than 2 fabrics results returned during test step 8

j-ororke committed:
Updating TC_ACL_2_7 python3 test module:
- Adding check to make sure that TH1 is empty binary data in response if TH2 is reading extensions unfiltered
- Added check to make sure that TH2 has empty binary data in response if TH1 is reading extension data unfiltered

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_7-8_fabrics-sensitive

cecille commented on src/python_testing/TC_ACL_2_8.py:
```diff
@@ -187,18 +259,28 @@ async def test_TC_ACL_2_8(self):
+        )
+        logging.info("TH1 read ACL result (fabric_filtered=False): %s", str(acl_list_unfiltered))
+        asserts.assert_equal(len(acl_list_unfiltered), 2, "Should have two ACL entries when not fabric filtered")
+
```
Can you please add a check here that the data for the non-accessing fabric is empty. You should get back a response, but it won't have any fields, and the client will fill in the defaults, which is empty everything. 

Please also add a comment as to why you're checking for empty everything.

cecille replied:
and wherever else this happens.

j-ororke committed:
Updating TC_ACL_2_8 test module:
- Adding asserts to verify that non-accessing fabric entry is empty or values are defaults to validate no data leak

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_7-8_fabrics-sensitive

j-ororke committed:
Update TC_ACL_2_8.py to resolve var name unknown after merge

- Updating events var for self._get_relevant_acl_events() from events to events_filtered due to new fabric filtered versus unfiltered logic

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_7-8_fabrics-sensitive

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_7-8_fabrics-sensitive

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_7-8_fabrics-sensitive

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_7-8_fabrics-sensitive

j-ororke committed:
Updating TC_ACL_2_7 and TC_ACL_2_8 test modules:
- Consolidated test harness controller validation of ACL data fields
- Updated functionality for assertions of checking extensions to being greater than 1 instead of equals 2
- Resolved linting errors

restyled-commits committed:
Restyled by autopep8

cecille reviewed (commented):
To check the pre-existing fabric condition - commission this with chip-tool first, open a commissioning window using chip-tool, then commission with the python test command line using the data from the chip-tool commissioning window.

j-ororke committed:
Update TC_ACL_2_7: Filter extension data by fabricIndex using list comprehension instead of static array positions per code review from Cecille

cecille committed:
Merge branch 'master' into update_TC_ACL_2_7-8_fabrics-sensitive
